onnxruntime/test/providers/provider_test_utils.cc

```diff
@@ -457,98 +486,115 @@ void OpTester::ExecuteModel(Model& model, InferenceSession& session_object, Expe
+              EXPECT_TRUE(inferred_dims.size() == expected_shape.NumDimensions());
+              for (size_t d = 0; d < inferred_dims.size(); ++d) {
+                // check equal unless the input involved a symbolic dimension
+                if (inferred_dims[d] != -1)
+                  EXPECT_EQ(expected_shape[d], inferred_dims[d]) << "Output idx = " << idx << " dim = " << d;
+              }
+            }
+            Check(expected_data, ort_value.Get<Tensor>(), provider_type);
+          } else {
+            Check(expected_data, ort_value, provider_type);
           }
+          ++idx;
+
+          // skip missing trailing optional outputs
+          if (idx == fetches.size())
+            break;
         }
-        Check(expected_data, ort_value.Get<Tensor>(), provider_type);
-      } else {
-        Check(expected_data, ort_value, provider_type);
       }
-      ++idx;
-
-      // skip missing trailing optional outputs
-      if (idx == fetches.size())
-        break;
     }
   }
+
+  return fetches;
 }
 
 void OpTester::Run(ExpectResult expect_result,
                    const std::string& expected_failure_string,
                    const std::unordered_set<std::string>& excluded_provider_types,
                    const RunOptions* run_options,
                    std::vector<std::unique_ptr<IExecutionProvider>>* execution_providers,
-                   ExecutionMode execution_mode) {
+                   ExecutionMode execution_mode,
+                   const CustomOutputVerifierFn& custom_output_verifier) {
   SessionOptions so;
   so.session_logid = op_;
   so.session_log_verbosity_level = 1;
   so.execution_mode = execution_mode;
   so.graph_optimization_level = TransformerLevel::Default;  // 'Default' == off
-  Run(so, expect_result, expected_failure_string, excluded_provider_types, run_options, execution_providers);
+  Run(so, expect_result, expected_failure_string, excluded_provider_types, run_options, execution_providers,
+      custom_output_verifier);
 }
 
 void OpTester::Run(SessionOptions so,  // Take the SessionOptions by value (i.e. make a copy) because we may need to modify it
                    ExpectResult expect_result,
                    const std::string& expected_failure_string,
                    const std::unordered_set<std::string>& excluded_provider_types,
                    const RunOptions* run_options,
-                   std::vector<std::unique_ptr<IExecutionProvider>>* execution_providers) {
+                   std::vector<std::unique_ptr<IExecutionProvider>>* execution_providers,
+                   const CustomOutputVerifierFn& custom_output_verifier) {
   std::string cur_provider = "not set";
   try {
 #ifndef NDEBUG
     run_called_ = true;
 #endif
-    auto p_model = BuildGraph();
+    fetches_.clear();
+    bool cacheEnabled = cached_model_ != nullptr;
+    auto p_model = !cacheEnabled ? BuildGraph() : cached_model_;
     auto& graph = p_model->MainGraph();
```

Comment:
Is p_model deduced to be a shared_ptr?

Reply:
correct, in future - we prefer to use unique_ptr for cached one, and return plain ptr.